src/gui/qgsmaptip.cpp

```diff
@@ -212,12 +213,30 @@ QString QgsMapTip::fetchFeature( QgsMapLayer *layer, QgsPointXY &mapPosition, Qg
     exp.prepare( &context );
     request.setSubsetOfAttributes( exp.referencedColumns(), vlayer->fields() );
   }
+
+  QgsRenderContext renderCtx = QgsRenderContext::fromMapSettings( mapCanvas->mapSettings() );
+  renderCtx.expressionContext() << QgsExpressionContextUtils::layerScope( vlayer );
+
+  std::unique_ptr< QgsFeatureRenderer > renderer;
+  if ( vlayer->renderer() )
+  {
+    renderer.reset( vlayer->renderer()->clone() );
+    renderer->startRender( renderCtx, vlayer->fields() );
+  }
+
   QgsFeatureIterator it = vlayer->getFeatures( request );
```

Comment:
It's possible to optimise this request based on the renderer's visible features -- if the layer has a renderer, you can use QgsFeatureRenderer::filter to set a filter expression on the request to avoid fetching and testing some invisible features, which should speed this functionality up

Reply:
Indeed! I take a look. Thanks for your input Nyall :+1:.

Reply:
@nyalldawson  I think I took into account your comment. Let me know if it's not what you had in mind.

Reply:
Mmm... In my use case with a graduated renderer and a SHP, the `renderer->filter( vlayer->fields() )` call returns an empty string so I get an "empty" iterator from `vlayer->getFeatures( request )`. So map tips are never displayed.

I can write a unit test to highlight this behavior if you want?

Reply:
It's not implemented for the graduated renderer -- so if filter() returns an empty string, you don't set a filter expression on the iterator. It's implemented for categorized, rule based renderers only.

Reply:
OK good to know. The code is now updated.